hpx/lcos/local/detail/condition_variable.hpp

```diff
@@ -115,20 +120,36 @@ namespace hpx { namespace lcos { namespace local { namespace detail
                 queue_.front().id_ = threads::invalid_thread_id_repr;
                 queue_.pop_front();
 
-                util::scoped_unlock<Lock> unlock(lock);
+                if (!queue_.empty())
+                {
+                    util::scoped_unlock<Lock> unlock(lock);
 
-                threads::set_thread_state(threads::thread_id_type(
-                    reinterpret_cast<threads::thread_data_base*>(id)),
-                    threads::pending, threads::wait_timeout,
-                    threads::thread_priority_default, ec);
-                if (!ec) return true;
+                    threads::set_thread_state(threads::thread_id_type(
+                        reinterpret_cast<threads::thread_data_base*>(id)),
+                        threads::pending, threads::wait_timeout,
+                        threads::thread_priority_default, ec);
+                    if (!ec) return true;
+                }
+                else
+                {
+                    // Since this is the last thread waiting on this condition
+                    // variable it could happen that this instance will be
+                    // destroyed before set_thread_state() returns. We have to
+                    // make sure that this instance is not touched anymore.
+                    lock.unlock();
```

Comment:
Wouldn't this change mean that we have to go over all uses of `notify_one` and make sure that the result is the expected one? I believe this would lead to `unlock()` being called in an already unlocked mutex.